curvefs/src/metaserver/s3compact_manager.cpp

```diff
@@ -183,16 +183,26 @@ void S3CompactManager::Stop() {
 }
 
 void S3CompactManager::Enqueue() {
-    std::vector<std::shared_ptr<S3Compact>> s3compacts;
+    std::vector<std::weak_ptr<S3Compact>> s3compacts;
     {
         // make a snapshot
         ReadLockGuard l(rwLock_);
         s3compacts = s3compacts_;
     }
     if (s3compacts.empty()) {
+        // fastpath
         sleeper_.wait_for(std::chrono::milliseconds(opts_.enqueueSleepMS));
+        return;
     }
-    for (const auto& s3compact : s3compacts) {
+    std::vector<int> toErase;
+    for (int i = 0; i < s3compacts.size(); i++) {
+        auto s3compact(s3compacts[i].lock());
+        if (!s3compact) {
```

Comment:
how to judge partition is deleted

Reply:
when partition is under deleting, the compact record should be deleted

Reply:
when the partition is deleting, it will reject update inode request.

Reply:
only partition hold the shared_ptr of s3compact, if partition is deleted, s3compact will be deleted.
weak_ptr lock() is thread safe.
if s3compact in partition has been deconstructed. lock() will return nullptr.
there is no deleting state, just deleted or not deleted.